vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java

```diff
@@ -201,24 +250,29 @@ private int buildElement(ArrayBackedValueStorage abvsFileNode, int fieldNum) thr
         int lastDot = contents.lastIndexOf('.');
         String type = contents.substring(lastDot + 1);
         String lastBit = contents.substring(firstColon + 1, lastDot);
-
+        int dots = contents.indexOf(".");
+        int nextdot;
+        String element="";
+        int elements=0;
         if (this.firstElement) {
             this.firstElement = false;
-            firstFinish = whereIFinish - this.childSequenceTypes.size() + 1;
-            String firstBit = contents.substring(1, firstColon);
-            List<String> names = new ArrayList<>();
-            List<String> values = new ArrayList<>();
-            List<String> uris = new ArrayList<>();
-            List<String> localNames = new ArrayList<>();
-            List<String> types = new ArrayList<>();
-            List<String> qNames = new ArrayList<>();
-            firstFinish = findAttributeChildren(firstFinish, names, values, uris, localNames, types, qNames);
-            Attributes atts = new IndexAttributes(names, values, uris, localNames, types, qNames);
-
-            handler.startElement(uri, firstBit, firstBit, atts);
-            buildElement(abvsFileNode, firstFinish + 1);
-            handler.endElement(uri, firstBit, firstBit);
-
+            while (dots <firstColon) {
+                nextdot = dots;
+                dots = contents.indexOf(".", dots + 1);
+                element = contents.substring(nextdot + 1, dots);
+                if (dots > firstColon) {
+                    element = contents.substring(nextdot + 1, firstColon);
+                }
+                List<String> names = new ArrayList<>();
```

Comment:
We should be able to run this code by creating a single names array and just clear with each pass. I think the correct way to fix this is to add the arrays to the IndexAttributes object with a reset function. Then the whole class could have one Attributes object that is reset before each use. Less object creation and more object reuse.